include/aos/common/crypto.hpp

```diff
@@ -0,0 +1,266 @@
+ */
+class Base64Encoder {
+public:
+    /**
+     * Encodes byte array to base64 string.
+     *
+     * @param blob input array data.
+     * @param[out] result  base64 encoded result string.
+     * @return void.
+     *
+     * TODO: add implementation
+     */
+    static void Encode(const Array<uint8_t>& /*data*/, String& /*result*/) { }
+};
+
+/**
+ * General certificate private key type.
+ */
+class PrivateKey {
+public:
+    /**
+     * Returns public part of a private key.
+     */
+    const Array<uint8_t>& GetPublic() const { return mPublicKey; }
+
+    /**
+     * Returns private part of a private key.
+     */
+    const Array<uint8_t>& GetPrivate() const { return mPublicKey; }
+
+private:
+    StaticArray<uint8_t, cCertificatePubKeyLen>  mPublicKey;
+    StaticArray<uint8_t, cCertificatePrivKeyLen> mPrivateKey;
+};
+
+namespace asn1 {
+/**
+ * ASN.1 OBJECT IDENTIFIER
+ */
+using ObjectIdentifier = StaticString<cASN1ObjIdLen>;
+
+/**
+ * ASN.1 structure extension. RFC 580, section 4.2
+ */
+struct Extension {
+    ObjectIdentifier                        mId;
+    StaticArray<uint8_t, cCertExtValueSize> mValue;
+};
+
+/**
+ * ASN1 Encodes string into ASN1 value
+ */
+
+inline void Encode(const String& /*src*/, Array<uint8_t>& /*asn1Value*/)
+{
+}
+
+/**
+ *
+ */
+
+} // namespace asn1
+
+namespace x509 {
+
+/**
+ * x509 Certificate.
+ */
+struct Certificate {
+    /**
+     * Certificate subject.
+     */
+    StaticString<cCertificateSubjSize> mSubject;
+
+    /**
+     * Certificate subject key id.
+     */
+    StaticArray<uint8_t, cCertKeyIdSize> mSubjectKeyId;
+
+    /**
+     * Certificate authority key id.
+     */
+    StaticArray<uint8_t, cCertKeyIdSize> mAuthorityKeyId;
+
+    /**
+     * Certificate issuer name.
+     */
+    StaticString<cCertificateIssuerLen> mIssuer;
+    /**
+     * Certificate serial number.
+     */
+    StaticArray<uint8_t, cCertificateSerialNumberLen> mSerial;
+    /**
+     * Certificate validity period.
+     */
+    time::Time mNotBefore, mNotAfter;
+};
+
+/**
+ * Provides interface to manage certificates.
+ */
+class CertificateProviderItf {
+public:
+    /**
+     * Creates a new certificate based on a template.
+     *
+     * @param templ a pattern for a new certificate.
+     * @param parent a parent certificate in a certificate chain.
+     * @param pubKey public key.
+     * @param privKey private key.
+     * @param[out] resultCert result certificate in PEM format.
+     * @result Error.
+     */
+    virtual Error CreateCertificate(
+        const Certificate& templ, const Certificate& parent, const PrivateKey& privKey, Array<uint8_t>& pemCert)
+        = 0;
+
+    /**
+     * Creates certificates from a PEM blob.
+     *
+     * @param pemBlob raw certificate in a PEM format.
+     * @param[out] resultCerts result certificates.
+     * @result Error.
+     */
+    virtual Error CreateCertificatesFromPEM(const Array<uint8_t>& pemBlob, Array<Certificate>& resultCerts) = 0;
```

Comment:
It actually, doesn't create it converts. So, `PEMToX509Cert` and `X509CertToPEM`  are good names here.